### PR TITLE
Implemented a method to log images

### PR DESCRIPTION
**Pull request by Aayush1104**

### CODE REVIEW

File: modules/logger/logger.py [hunk 1]
```diff
@@ -13,6 +13,8 @@
 # pylint: disable-next=unused-import
 import types
```
Comment (maxlou05) [inline modules/logger/logger.py:14]: I don't think this is used anywhere? You can probably remove it

File: modules/logger/logger.py [hunk 2]
```diff
@@ -13,6 +13,8 @@
...
 
+import numpy as np
+from PIL import Image
 from ..read_yaml import read_yaml
```
Comment (maxlou05) [inline modules/logger/logger.py:18]: Add a space above this line. We separate imports from our own files from imports of external libraries (btw, we also separate Python default libraries before those).

File: modules/logger/logger.py [hunk 3]
```diff
@@ -143,6 +145,48 @@ def info(self, message: str, log_with_frame_info: bool = True) -> None:
...
+            self.error("No file handler found to save image.")
+            return
+
+        log_dir = os.path.dirname(file_handler.baseFilename)
+
+        img_dir = os.path.join(log_dir, "images")
+        os.makedirs(img_dir, exist_ok=True)
+
+        timestamp = datetime.datetime.now().strftime("%Y%m%d-%H%M%S")
+        img_path = os.path.join(img_dir, f"{filename}_{timestamp}.png")
```
Comment (maxlou05) [inline modules/logger/logger.py:180]: I feel like you don't need to do this, the logger already has a file handler initialized in create. You just need to save an image.

File: modules/logger/logger.py [hunk 4]
```diff
@@ -143,6 +145,48 @@ def info(self, message: str, log_with_frame_info: bool = True) -> None:
...
+        img.save(img_path)
```
Comment (maxlou05) [inline modules/logger/logger.py:181]: So just save the original filename that's passed in, ie `img.save(filename)`

File: modules/logger/logger.py [hunk 5]
```diff
@@ -143,6 +145,48 @@ def info(self, message: str, log_with_frame_info: bool = True) -> None:
             message = self.message_and_metadata(message, caller_frame)
         self.logger.info(message)
 
+    def save_image(
```
Comment (maxlou05) [inline modules/logger/logger.py:148]: Also, can you move this function to the bottom (like after all the other log text things functions?)

**Review by maxlou05 — COMMENTED** [review]
Reviewed

File: modules/logger/logger.py [hunk 6]
```diff
@@ -11,7 +11,8 @@
 
 # Used in type annotation of logger parameters
 # pylint: disable-next=unused-import
-import types
```
Comment (maxlou05) [inline modules/logger/logger.py:14]: My bad, I just realized this is used elsewhere, can you bring it back? (Also add a space between this and import numpy)

File: modules/logger/logger.py [hunk 7]
```diff
@@ -172,3 +173,25 @@ def critical(self, message: str, log_with_frame_info: bool = True) -> None:
             caller_frame = logger_frame.f_back
             message = self.message_and_metadata(message, caller_frame)
         self.logger.critical(message)
+
+    def save_image(
+        self, image: np.ndarray, filename: str, log_with_frame_info: bool = True
```
Comment (maxlou05) [inline modules/logger/logger.py:178]: Can you add an optional parameter: log_info_message with default to false, and only log a message if this is true?

**Review by maxlou05 — COMMENTED** [review]
Reviewed

File: modules/logger/logger.py (hunk 7 above)
Comment (Aayush1104) [reply modules/logger/logger.py:178]: would log_with_frame_info's part stay the same? I'd only had an if condition for line 199 correct?

**Review by maxlou05 — APPROVED** [review]
Approved!

**Pull request merged**